File: .kotlin_alpha/services/s3/src/main/kotlin/com/kotlin/s3/DeleteBucket.kt
//snippet-sourcedescription:[DeleteBucket.kt demonstrates how to delete an Amazon Simple Storage Service (Amazon S3) bucket.]
//snippet-keyword:[AWS SDK for Kotlin]
//snippet-keyword:[Code Sample]
//snippet-service:[Amazon S3]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[05/06/2021]
//snippet-sourceauthor:[scmacdon-aws]

/*
   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   SPDX-License-Identifier: Apache-2.0
*/

package com.kotlin.s3

// snippet-start:[s3.kotlin.del_bucket.import]
import aws.sdk.kotlin.services.s3.S3Client
import aws.sdk.kotlin.services.s3.model.DeleteBucketRequest
import aws.sdk.kotlin.services.s3.model.S3Exception
import kotlin.system.exitProcess
// snippet-end:[s3.kotlin.del_bucket.import]

/**
To run this Kotlin code example, ensure that you have setup your development environment,
including your credentials.

For information, see this documentation topic:
https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/setup.html
 */
suspend fun main(args: Array<String>) {

    val usage = """
    Usage:
        <bucketName> 

    Where:
        bucketName - the name of the Amazon S3 bucket to delete.
    """

    if (args.size != 1) {
        println(usage)
        exitProcess(0)
     }

    val bucketName = args[0]
    val s3Client = S3Client { region = "us-east-1" }
    deleteExistingBucket(s3Client,bucketName)
    s3Client.close()
}

// snippet-start:[s3.kotlin.del_bucket.main]
 suspend fun deleteExistingBucket(s3Client: S3Client, bucketName: String?) {

        try {
            val deleteBucketRequest = DeleteBucketRequest {
                bucket = bucketName
            }
            s3Client.deleteBucket(deleteBucketRequest)
            println("The $bucketName was successfully deleted!")

        } catch (e: S3Exception) {
            println(e.message)
            s3Client.close()
            exitProcess(0)
        }
    }
// snippet-end:[s3.kotlin.del_bucket.main]

File: cpp/example_code/s3_copy/delete_bucket.cpp
//snippet-sourcedescription:[delete_bucket.cpp demonstrates how to delete an Amazon Simple Storage Service (Amazon S3) bucket.]
//snippet-keyword:[AWS SDK for C++]
//snippet-keyword:[Code Sample]
//snippet-service:[Amazon S3]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[12/15/2021]
//snippet-sourceauthor:[scmacdon - aws]

/*
   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   SPDX-License-Identifier: Apache-2.0
*/

// snippet-start:[s3.cpp.delete_bucket.inc]
#include <iostream>
#include <aws/core/Aws.h>
#include <aws/s3/S3Client.h>
#include <aws/s3/model/DeleteBucketRequest.h>
#include "awsdoc/s3/s3_examples.h"
// snippet-end:[s3.cpp.delete_bucket.inc]

/* 
 * 
 * Prerequisites: The bucket to be deleted.
 *
 * Inputs:
 * - bucketName: The name of the bucket to delete.
 * - region: The AWS Region of the bucket to delete.
 * 
 *  To run this C++ code example, ensure that you have setup your development environment, including your credentials.
 *  For information, see this documentation topic:
 *  https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/getting-started.html
 * 
 */

 // snippet-start:[s3.cpp.delete_bucket.code]
using namespace Aws;

bool AwsDoc::S3::DeleteBucket(const Aws::String &bucketName, const Aws::String &region)
{
    Aws::Client::ClientConfiguration clientConfig;
    if (!region.empty()) {
        clientConfig.region = region;
    }

    Aws::S3::S3Client client(clientConfig);

    Aws::S3::Model::DeleteBucketRequest request;
    request.SetBucket(bucketName);

    Aws::S3::Model::DeleteBucketOutcome outcome =
            client.DeleteBucket(request);

    if (!outcome.IsSuccess())
    {
        auto err = outcome.GetError();
        std::cout << "Error: DeleteBucket: " <<
                  err.GetExceptionName() << ": " << err.GetMessage() << std::endl;
        return false;
    }
    else
    {
        std::cout << "The bucket was deleted" << std::endl;
        return true;

    }
}

int main()
{
    //TODO: Change bucket_name to the name of a bucket in your account.
    //If the bucket is not in your account, you will get one of two errors:
    Aws::String bucketName = "<Bucket Name>";
    //TODO:  Set to the AWS Region of the bucket bucket_name.
    Aws::String region = "us-east-1";

    Aws::SDKOptions options;
    Aws::InitAPI(options);

    AwsDoc::S3::DeleteBucket(bucketName, region);

    ShutdownAPI(options);
 }
// snippet-end:[s3.cpp.delete_bucket.code]
